lib/proxy_api/external_ipam.rb

```diff
@@ -0,0 +1,198 @@
+    #
+    # Returns: An array of sections on success, or a hash with "error" otherwise.
+    #
+    # Examples
+    #   Response if success: [
+    #     {"id":"4","name":"Test Section","description":"A Test Section"},
+    #     {"id":"5","name":"Awesome Section","description":"A totally awesome Section"}
+    #   ]
+    #   Response if no sections exist:
+    #     {"code":200, "success":true, "message": "No sections available", "time": 0.004, "data": []}
+    #   Response if can't connect to External IPAM server
+    #     {"error": "Unable to connect to External IPAM server"}
+    def get_sections
+      parse get("/sections")
+    rescue => e
+      raise ProxyException.new(url, e, N_("Unable to obtain sections from External IPAM."))
+    end
+
+    # Get a list of subnets for the given External IPAM section.
+    #
+    # Params:  1. section_name: The name of the External IPAM section
+    #
+    # Returns: An array of subnets on success, or a hash with "error" otherwise.
+    #
+    # Examples:
+    #   Response if success: [
+    #     {"id":"9","subnet":"100.10.10.0","mask":"24","sectionId":"5","description":"Test Subnet 1"},
+    #     {"id":"10","subnet":"100.20.20.0","mask":"24","sectionId":"5","description":"Test Subnet 2"}
+    #   ]
+    #   Response if section not found:
+    #     {"error": "Section not found in External IPAM."}
+    #   Response if no subnets exist in section.
+    #     {"code": 200, "success": true, "data": [], "time": 0.007}
+    #   Response if can't connect to External IPAM server
+    #     {"error": "Unable to connect to External IPAM server"}
+    def get_subnets_by_section(section_name)
+      raise "section_name must be provided" if section_name.blank?
+      parse get("/sections/#{URI.encode(section_name)}/subnets")
+    rescue => e
+      raise ProxyException.new(url, e, N_("Unable to obtain subnets in section/group %s from External IPAM."), section_name)
+    end
+
+    # Returns an array of subnets from External IPAM. If the "Require unique subnets" setting in External IPAM
+    # is enabled, this will return one subnet(or zero if not exists). If this setting is disabled(i.e. you can
+    # have duplicate subnets and IP's), then it is possible to get more than one.
+    #
+    # Params:  1. subnet: The Foreman subnet object
+    #
+    # Returns: A subnet with "data" key, or a hash with "error" otherwise
+    #
+    # Examples:
+    #   Response if subnet(s) exists:
+    #     {"code": 200, "success": true, "data": [
+    #       {"id": "32", "subnet": "10.10.10.0", "description": "Test", "mask": "29"}
+    #     ], "time": 0.007}
+    #   Response if subnet not exists:
+    #     {"error": "No subnets found"}
+    #   Response if can't connect to External IPAM server
+    #     {"error": "Unable to connect to External IPAM server"}
+    def get_subnet(subnet)
+      raise "subnet.network cannot be nil" unless subnet.network.present?
+      parse get("/get_subnet?cidr=#{subnet.network_address}")
```

Comment:
This needs to be reworked - input must be subnet address not subnet Foreman instance (4vs6 problem).